text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1875 @@
+create table student(
+    id int primary key,
+    name text
+)
+```
+
+When the string is evaluated, the leading indentation is removed. It is **dedented**.
+
+In the source code, the string is **indented**.
+
+- When viewing the string from the source code, the indentation is obvious.
+
+  However, it is *not* obvious what will happen to the string when it is evaluated. "Dedent" can be clearer in this regard, as we already have 1 piece of information and the word "dedent" brings us the other piece.
+
+- The string may not always be considered to be indented:
+
+   ```rs
+   let _ = d"
+   hello world
+   ";
+   ```
+
+  In the above example, there is no indentation for the strings. It would be inaccurate to describe the string as having indentation.
+
+  Once the string is evaluated, it is accurate to describe the removal of the non-existing indentation as still "dedenting" the string.
+
+#### Why not `m` for "multi-line"
+
+- Dedented string literals do not necesserily represent a multi-line string:
+
+```rs
+let _ = d"
+hello world
+";
+```
+
+The above is equivalent to:
+
+```rs
+let _ = "hello world";
+```
+
+Confusion could arise, as people expect it to evaluate to a string spanning multile lines.
+
+#### Why not `h` for "heredoc"
+
+RFC #3450 uses `h` as the modifier instead of `d`, as an acronym for [Here document](https://en.wikipedia.org/wiki/Here_document).
+
+- The term is likely to be less known, and may raise confusion, especially amongst
+  those that don't know what it is.
+- Here documents are more associated with "code blocks", which may associate an "info string"
+  with them (such as in markdown). This RFC does not propose an info string.
+
+While the feature this RFC proposes (dedented string literals) are useful for code
+blocks, it is not just for them.
+
+### The choice of the form `d"string"`
+
+The syntax of `d"string"` is chosen for the following reasons:
+
+- Fits with existing string modifiers, such as `b"string"`, `r#"string"#"` and `c"string"`
+- Composes with existing string modifiers: `db"string"`, `dc"string"`, `dr#"string"#`, and `dbr#"string"#`. 
+- Does not introduce a lot of new syntax. Dedented string literals can be explained in terms of existing language features.
+- Adding a single letter `d` before a string literal to turn it into a dedented string literal is an incredibly easy modification.
+- Rust reserves space for additional string modifiers.
+
+  Adding this feature does not require a new edition, as it is backwards-compatible for editions later than Edition 2024, as the syntax has been [reserved](https://doc.rust-lang.org/edition-guide/rust-2024/reserved-syntax.html) since this edition.
+
+The choice for `d` to come before all other modifiers is not arbitrary.
+
+Consider `dbr` and all possible alternatives:
+
+1. `dbr`: dedented byte raw string
+1. `bdr`: byte dedented raw string
+1. `brd`: byte raw dedented string
+
+The first example reads in the most natural manner. The other two don't.
```

Comment:
`d` and `r` change how you process a string while `b` and `c` change the type.  I feel like `d` and `r` should be grouped with each other.

Thinking over https://github.com/rust-lang/rfcs/pull/3830/files#r2136431112, I think a natural way of modeling this is operations being performed
- `d` or `r` should come *before* `b` or `c` (is "before" on the left or right? I lean towards right like function calls / bind most tightly) as that is processed before things related to the type, like adding a null terminator
- Unsure of implications for having `d` or `r` coming first

So I think there is a serious reason for us to consider `bdr` or `brd`.  I lean towards `brd` as my mind for some reason feels like describing it as dedenting before treating it as raw.

Reply:
Another argument in favor of `[bc]?d?r?` (type, content_modification, string_delimiter_and_escapes):
The output type seems way more important to the one reviewing the code than the indentation mode/level.

> Unsure of implications for having d or r coming first

One argument against `brd`:
`br###d"Hello World"###`
`bdr###"Hello World"###`
It is visually harder to to see what is part of the string content. It's less noticeable with `rd"Hello World"` because it doesn't have the `#` delimiter, which currently behaves like brackets. At least in my mind I see `r` more closely linked to when the string ends than to whether escape sequences are parsed, but that's probably just because I used it for that more often.